Initial implementation

mcdonnnj opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request creates a [Terraform root module] based on the [`basic_usage` example in cisagov/skeleton-tf-module](https://github.com/cisagov/skeleton-tf-module/tree/d27789b21f87d9d99bf2cbd2ce1093a205105327/examples/basic_usage).
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

We have a large number of [Terraform root module] projects that we maintain and it makes sense to have a derivative skeleton to reflect the specific needs of a root module versus a reusable module.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass.

> [!NOTE]
> We should figure out where to do a test deploy to ensure the configuration will successfully `terraform apply`.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release (necessary if and only if the version was bumped).

[Terraform root module]: https://developer.hashicorp.com/terraform/language/modules#the-root-module

mcdonnnj committed:
Remove Terraform module content

Since this will be a root module we do not need the reusable module
content.

mcdonnnj committed:
Move the basic usage example to the root of the project

The existing basic usage example for a Terraform module should be
suitable as an example Terraform root module with some tweaks.

mcdonnnj committed:
Update the source for the `example` module

Since this project does not house the `example` module we must update
the source to point to the cisagov/skeleton-tf-module repository on
GitHub. We also adjust the module's meta arguments to be sorted
alphabetically in keeping with our best practices.

mcdonnnj committed:
Add a backend configuration

Since this project represents a Terraform root module we will need a
backend to store Terraform state.

mcdonnnj committed:
Remove the examples/ directory

This directory is no longer relevant with the flattening of its
contents to the repository's root.

mcdonnnj committed:
Move the provider configuration into its own file

This fits our standard practice.

mcdonnnj committed:
Update the Terraform information in the README

This is done automatically using `terraform-docs`.

mcdonnnj committed:
Update the dependabot configuration

Enable the ignore directive for the root directory and remove the
configuration for the examples/basic_usage/ directory that was itself
removed.

mcdonnnj committed:
Update the README

Update the README's language to reflect the purpose of the repository.
Run `terraform-docs` (using my fork at mcdonnnj/terraform-docs) to
update the information in the README.

mcdonnnj committed:
Use a specific version of cisagov/skeleton-tf-module

It is a best practice to use version pinning when possible.

mcdonnnj commented on main.tf:
```diff
@@ -1,61 +1,14 @@
+  ami_owner_account_id  = var.ami_owner_account_id
+  aws_availability_zone = var.aws_availability_zone
+  aws_region            = var.aws_region
+  subnet_id             = aws_subnet.example.id
```
My friend it is right here:
https://github.com/cisagov/skeleton-tf-root-module/blob/7cfe57156cd223430eb964f699be0b700f21d0ca/vpc.tf#L15-L20

dav3r commented on backend.tf:
```diff
@@ -0,0 +1,10 @@
+terraform {
+  backend "s3" {
+    bucket         = "cisa-cool-terraform-state"
```
I'd like to see this skeleton follow the path laid out by PR's such as https://github.com/cisagov/cool-accounts/pull/161 and [all of these other PRs](https://github.com/search?q=org%3Acisagov+%22Support+new+COOL+environment+structure%22&type=pullrequests):
```suggestion
    # Use a partial configuration to avoid hardcoding the bucket name. This
    # allows the bucket name to be set on a per-environment basis via the
    # -backend-config command line option or other methods.  For details, see:
    # https://developer.hashicorp.com/terraform/language/backend#partial-configuration
    bucket         = ""
```

Note that the README should also be updated to include the instructions on how to configure the backend.  Unfortunately, I did not think to put those updated instructions in every PR linked above, but you can refer to https://github.com/cisagov/skeleton-packer/pull/393 for a good example to copy.

mcdonnnj replied:
Please see [cb0e8f2](https://github.com/cisagov/skeleton-tf-root-module/pull/39/commits/cb0e8f22449d33c9617d71f537b5cb911623ee41) and let me know what you think.

dav3r replied:
Looks great- all that I think is missing is a commit like https://github.com/cisagov/publish-egress-ip-terraform/pull/49/commits/60b2c5d1e88407ddfa21c951c4bcaefdef93eed6 to add `*.tfconfig` to the `.gitignore` file.

mcdonnnj replied:
Ah, right, forgot that part. Please see [98e0e68](https://github.com/cisagov/skeleton-tf-root-module/pull/39/commits/98e0e685f5a8f64a228bd5d88bc2130f06aa8528).

dav3r replied:
Perfect, thanks!

dav3r commented on .github/dependabot.yml:
```diff
@@ -31,17 +31,9 @@ updates:
-    # ignore:
-    #   # Managed by cisagov/skeleton-tf-module
-    #   - dependency-name: hashicorp/aws
+    ignore:
+      # Managed by cisagov/skeleton-tf-module
+      - dependency-name: hashicorp/aws
     package-ecosystem: terraform
```
Is this skeleton meant to be a (lineage) descendant of `skeleton-tf-module`?  If so, you are missing the `.github/lineage.yml` file.  If not, then I don't think we should be ignoring this dependency here.

mcdonnnj replied:
Updating the `.github/lineage.yml` file is [handled by the `gh skeleton` command](https://github.com/cisagov/gh-skeleton/blob/4c72df155535a03220ea90641b616db409b57f3d/gh-skeleton#L174-L181) and you can see it is already updated:
https://github.com/cisagov/skeleton-tf-root-module/blob/42371841521fe34af834934fc8f02a468308b4d2/.github/lineage.yml#L1-L5

dav3r replied:
Aha, I forgot about that; I was only looking at the code changes in the PR. 👍

dav3r commented on providers.tf:
```diff
@@ -1,13 +1,11 @@
-#   region = var.aws_region
-# }
+provider "aws" {
+  # Our primary provider uses our terraform role
```
Capitalization:
```suggestion
  # Our primary provider uses our Terraform role
```

mcdonnnj replied:
This should probably get punted up to [cisagov/skeleton-tf-module](https://github.com/cisagov/skeleton-tf-module/blob/d27789b21f87d9d99bf2cbd2ce1093a205105327/examples/basic_usage/main.tf#L2) because this project just pulled the provider configuration from the `basic_usage` example.

dav3r replied:
I made https://github.com/cisagov/skeleton-tf-module/pull/249 per your suggestion, but you may as well fix it here on this PR too.

mcdonnnj replied:
Please see [14d0a1f](https://github.com/cisagov/skeleton-tf-root-module/pull/39/commits/14d0a1f9ac201598e8169f806ebfe74690743209). This may get clobbered out of existence when I rebase after the incoming lineage PR is merged.

dav3r reviewed (changes requested):
Looks good overall, but I've got a few items for you to consider.

mcdonnnj committed:
Use a partial backend configuration

We have moved to using a partial backend configuration to both avoid
hardcoding S3 bucket names in our Terraform configurations and to make
it easier to use the same configuration across multiple, discrete
deployments of that configuration.

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

mcdonnnj committed:
Add `*.tfconfig` files to the `.gitignore` configuration

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

mcdonnnj commented on main.tf:
```diff
@@ -1,61 +1,14 @@
+  ami_owner_account_id  = var.ami_owner_account_id
+  aws_availability_zone = var.aws_availability_zone
+  aws_region            = var.aws_region
+  subnet_id             = aws_subnet.example.id
```
Yes, it still exists, but thank you for trying.

mcdonnnj commented on backend.tf:
```diff
@@ -0,0 +1,14 @@
+    bucket         = ""
+    dynamodb_table = "terraform-state-lock"
+    encrypt        = true
+    key            = "skeleton-tf-root-module/terraform.tfstate"
```
Downstream projects should update the key to use their own project name. I don't think there's much value to be had in moving this into a tfconfig file as part of the partial configuration like we do for the bucket. @cisagov/vm-dev thoughts?

dav3r replied:
We don't do this in any of our other skeletons and it hasn't really been an issue, so I'm good with it as is.

mcdonnnj committed:
Update the `bump-version` list of version files to update

Remove the README because it does not reference the project's version
after the conversion to a root module.

mcdonnnj committed:
Reset the version for an initial release

The initial release of this project should be v1.0.0.

mcdonnnj commented on main.tf:
```diff
@@ -1,73 +1,14 @@
+  ami_owner_account_id  = var.ami_owner_account_id
+  aws_availability_zone = var.aws_availability_zone
+  aws_region            = var.aws_region
+  subnet_id             = aws_subnet.example.id
```
https://github.com/cisagov/skeleton-tf-root-module/pull/39#discussion_r2266841178

mcdonnnj commented on main.tf:
```diff
@@ -1,73 +1,14 @@
+#-------------------------------------------------------------------------------
+# Configure the example module.
+#-------------------------------------------------------------------------------
+module "example" {
+  providers = {
+    aws = aws
+  }
+  source = "github.com/cisagov/skeleton-tf-module?ref=v1.1.0"
```
@cisagov/vm-dev Open to opinions on this one.

jsf9k replied:
The version is for cisagov/skeleton-tf-module, not cisagov/skeleton-tf-root-module.  Is this not a red herring?

mcdonnnj replied:
Yes, I'm just confirming that we have no reason or desire to lockstep the versions.

jsf9k replied:
Correct, I don't think there is any reason or desire to lockstep the versions.

dav3r reviewed (approved):
👍 👍
